### PR TITLE
report: fix category anchor links for embedded render

### PR DESCRIPTION
**Pull request by connorjclark**
This link was replacing the URL with basically `window.location.pathname#...`, which is a bug when the report is not on the root of an origin. bug was made apparent in new PSI.

### CODE REVIEW

File: report/renderer/category-renderer.js [hunk 1]
```diff
@@ -319,7 +319,11 @@ export class CategoryRenderer {
   renderScoreGauge(category, groupDefinitions) { // eslint-disable-line no-unused-vars
     const tmpl = this.dom.createComponent('gauge');
     const wrapper = this.dom.find('a.lh-gauge__wrapper', tmpl);
-    this.dom.safelySetHref(wrapper, `#${category.id}`);
+    const url = new URL(window.location.href);
+    url.hash = category.id;
+    this.dom.safelySetHref(wrapper, url.href);
+    // wrapper.href = url.href;
```
Comment (connorjclark) [inline report/renderer/category-renderer.js:325]: if the modification in `safelySetHref` is not desirable, we could do one of these ... thoughts ?

File: report/renderer/report-ui-features.js [hunk 2]
```diff
@@ -128,6 +128,7 @@ export class ReportUIFeatures {
       const containerEl = this._dom.find('.lh-container', this._document);
       const elToAddScrollListener = this._getScrollParent(containerEl);
       elToAddScrollListener.addEventListener('scroll', this._updateStickyHeaderOnScroll);
+      window.addEventListener('hashchange', this._updateStickyHeaderOnScroll);
```
Comment (connorjclark) [inline report/renderer/report-ui-features.js:131]: drive by imporvement!

File: report/renderer/dom.js [hunk 3]
```diff
@@ -157,7 +157,7 @@ export class DOM {
       return;
     }
 
-    const allowedProtocols = ['https:', 'http:'];
+    const allowedProtocols = ['https:', 'http:', window.location.protocol];
```
Comment (adamraine) [inline report/renderer/dom.js:160]: Is this mostly to account for `file:` protocols? Could we ignore `allowedProtocols` if the origin matches `window.location` instead?

Comment (connorjclark) [reply report/renderer/dom.js:160]: Yes, it was to account for `file:` when viewing a standalone report.

> Could we ignore allowedProtocols if the origin matches window.location instead?

Isn't this the same behavior?

Comment (adamraine) [reply report/renderer/dom.js:160]: > Isn't this the same behavior?

I meant we could bypass `allowedProtocols` if the `url.origin` matches `location.origin`.

**Comment by brendankenny** [comment]
> This link was replacing the URL with basically window.location.pathname#..., which is a bug when the report is not on the root of an origin. bug was made apparent in new PSI.

can you give an example of what was broken? These are setting the `href` to `#category`...my understanding was clicking a fragment link like that takes you to `new URL('#category', window.location)`, which will then scroll to the fragment.

```js
const url = new URL(window.location.href);
url.hash = category.id;
this.dom.safelySetHref(wrapper, url.href);
```

does the same thing, so I think I'm missing what breaks with the plain fragment approach.

**Comment by connorjclark** [comment]
Oh, then I must not of understood what is broken from @exterkamp. Shane, can you explain?

**Comment by connorjclark** [comment]
something to do with the app router for new psi. closing.

**Comment by connorjclark** [comment]
We found that using a fully qualified link gets around the issue (and is in fact the prescribed solution for the internal framework being used).

File: report/renderer/category-renderer.js (hunk 1 above)
Comment (connorjclark) [reply report/renderer/category-renderer.js:325]: (TODO: add comment about why fully qualified link is needed)

Comment (connorjclark) [reply report/renderer/category-renderer.js:325]: fyi `wrapper.addEventListener('click', () => window.location.hash = category.id);` would work too ... but is that bad for a11y?

**Comment by brendankenny** [comment]
> We found that using a fully qualified link gets around the issue (and is in fact the prescribed solution for the internal framework being used).

can we define the issue, though? :D

**Comment by connorjclark** [comment]
Clicking the score gauges should move the page to that id element, but in the new psi/LB instead the page reloads to the root view. The framework is very hostile against any URL changes. The exact mechanism of the bug is unknown to me, but I found someone with the exact same issue in the internal QA site and the solution was to use the fully qualified URL.

**Comment by brendankenny** [comment]
> can we define the issue, though? :D

tl;dr: the embedder page always comes with a (more or less) `<base href="${window.location.origin}">`.

That's fine if you're in the root directory, but if you're on `https://example.com/lh/report/viewer.html`, a `<a href="#category">` takes you to `https://example.com#category`, _not_ `https://example.com/lh/report/viewer.html#category`.

:/

**Comment by brendankenny** [comment]
Options:

1. (current PR): set score gauges to `new URL(`#${category.id}`, window.location.href)`. Downsides:
   - have to remember everywhere that might create fragment links. Could change the `url.startsWith('#')` condition in `safelySetHref` to do this for all setting of fragments. That still won't catch if anyone adds a fragment link in `templates.html` at some point, though, but there aren't any currently.
   - "Save as HTML" in the plain report saves the outerHTML of itself. When opening the saved report, it will re-render and so get updated relative links, but this does mean that the former `window.location.href` of the report is saved in the html, potentially exposing local paths or URLs if originally hosted remotely. If someone opens up the report without JS (or has some html-pdf saving going on or whatever), they'll also get the old remote links instead of local ones.

2. click handler: `wrapper.addEventListener('click', () => window.location.href = category.id)`. Downsides:
   - we should avoid overriding default behavior when possible (let a link be a link). Arguably setting fully qualified links (option 1) is the semantically correct thing to do when we can't rely on the `<base>` url rather than setting a relative fragment url _and_ overriding the behavior on click anyways.

It's also worth bringing up a slight variant:

3. After `renderReport()`, have the embedder look for links and override them with option (1) (or option (2) if for some reason that helps something). e.g.
   ```js
   const reportEl = renderReport(lhr, container);
   // Work around <base href> by resolving relative links against the current page.
   reportEl.querySelectorAll('a[href^="#"]')
     .forEach(a => a.href = new URL(a.getAttribute('href'), window.location.href));
   ``` 
   this would catch all fragment links, added in JS or in our templates, but requires the embedder to remember to do it and it would have to use viewer-style "Save as HTML" (using the generator) instead of standalone-report style.

**Comment by connorjclark** [comment]
Since we have time before this becomes an issue in a published product, we will wait and see if we can get the bug fixed at the source. If not, our workarounds are quick enough to apply in a last-minute crunch :)

**Comment by connorjclark** [comment]
#13016
